### PR TITLE
Feature/with statement js compiler

### PR DESCRIPTION
**Pull request by TobiasWienand**
Addresses #437 
Note to reviewer: https://github.com/TobiasWienand/fuzzilli/blob/7ac6ae5fc777ed469413682640f33a3f7d9a40a3/Sources/Fuzzilli/Compiler/Compiler.swift#L487
needs attention. Is this the correct way to iterate over the block of the with-statement?

### CODE REVIEW

File: Sources/Fuzzilli/Compiler/Compiler.swift [hunk 1]
```diff
@@ -479,6 +479,16 @@ public class JavaScriptCompiler {
             let value = try compileExpression(throwStatement.argument)
             emit(ThrowException(), withInputs: [value])
 
+        case .withStatement(let withStatement):
+            let object = try compileExpression(withStatement.object)
+            let beginWith = BeginWith()
+            emit(beginWith, withInputs: [object])
```
Comment (saelo) [inline Sources/Fuzzilli/Compiler/Compiler.swift:485]: Nit: maybe combine into one line: `emit(BeginWith(), ...)`

File: Sources/Fuzzilli/Compiler/Compiler.swift [hunk 2]
```diff
@@ -479,6 +479,16 @@ public class JavaScriptCompiler {
...
+            try enterNewScope {
+                for statement in withStatement.body.blockStatement.body {
```
Comment (saelo) [inline Sources/Fuzzilli/Compiler/Compiler.swift:487]: I think instead of this, you can just do `try compileBody(withStatement.body)`. Does that work?

**Review by saelo — COMMENTED** [review]
Nice, thanks!

File: Sources/Fuzzilli/Compiler/Compiler.swift (hunk 2 above)
Comment (TobiasWienand) [reply Sources/Fuzzilli/Compiler/Compiler.swift:486]: That's what I tried initially, however VSCode says: 

> For-in loop requires 'Compiler_Protobuf_Statement' to conform to 'Sequence'

Below is an example js code using the with-operator and the AST generated from the JS parser
# basic.js
```
var person = {
    firstName: "John",
  };
  
with (person) {
    console.log(firstName);
  }
  
```
# AST of basic.js
```
Fuzzilli.Compiler_Protobuf_AST:
statements {
  variableDeclaration {
    declarations {
      name: "person"
      value {
        objectExpression {
          fields {
            property {
              name: "firstName"
              value {
                stringLiteral {
                  value: "John"
                }
              }
            }
          }
        }
      }
    }
  }
}
statements {
  withStatement {
    object {
      identifier {
        name: "person"
      }
    }
    body {
      blockStatement {
        body {
          expressionStatement {
            expression {
              callExpression {
                callee {
                  memberExpression {
                    object {
                      identifier {
                        name: "console"
                      }
                    }
                    name: "log"
                  }
                }
                arguments {
                  identifier {
                    name: "firstName"
                  }
                }
              }
            }
          }
        }
      }
    }
  }
}
```

Comment (saelo) [reply Sources/Fuzzilli/Compiler/Compiler.swift:486]: Hm which for-in loop does it complain about?
What I mean is something like
```swift
case .withStatement(let withStatement):
    let object = try compileExpression(withStatement.object)
    emit(BeginWith(), withInputs: [object])
    try enterNewScope {
        try compileBody(withStatement.body)
    }
    emit(EndWith())
```
`compileBody` requires only a `StatementNode`, so I think that should work there.

Comment (TobiasWienand) [reply Sources/Fuzzilli/Compiler/Compiler.swift]: I see. That works 👍

**Comment by saelo** [comment]
Thanks!

**Pull request merged**